File: frontend/real-estate-client/src/app/components/services/property.service.ts
import { Injectable } from '@angular/core';
import { Property } from '../models/property.model';

@Injectable({ providedIn: 'root' })
export class PropertyService {
  private properties: Property[] = [
    {
      id: 1,
      address: 'HaYarkon 99, Tel Aviv',
      pricePerNight: 620,
      isAvailable: true,
      imageUrl: 'https://example.com/images/property1.jpg',
      description: 'Spacious 2BR apartment near the beach.'
    },
    {
      id: 2,
      address: 'King George 17, Jerusalem',
      pricePerNight: 480,
      isAvailable: false,
      imageUrl: 'https://example.com/images/property2.jpg',
      description: 'Quiet 1BR apartment close to city center.'
    }
  ];

  getAll(): Property[] {
    return this.properties;
  }

  search(term: string): Property[] {
    if (!term) return this.properties;
    const lower = term.toLowerCase();
    return this.properties.filter(p =>
      p.address.toLowerCase().includes(lower) ||
      p.description.toLowerCase().includes(lower)
    );
  }

  filterAvailable(onlyAvailable: boolean): Property[] {
    return onlyAvailable
      ? this.properties.filter(p => p.isAvailable)
      : this.properties;
  }

  sortByPrice(descending: boolean): Property[] {
    return [...this.properties].sort((a, b) =>
      descending ? b.pricePerNight - a.pricePerNight : a.pricePerNight - b.pricePerNight
    );
  }

  query(term: string, onlyAvailable: boolean, descending: boolean): Property[] {
    let result = this.getAll();

    if (term)
      result = result.filter(p =>
        p.address.toLowerCase().includes(term.toLowerCase()) ||
        p.description.toLowerCase().includes(term.toLowerCase())
      );

    if (onlyAvailable)
      result = result.filter(p => p.isAvailable);

    result = result.sort((a, b) =>
      descending ? b.pricePerNight - a.pricePerNight : a.pricePerNight - b.pricePerNight
    );

    return result;
  }
}

File: backend/Services/PropertyService.cs
// PropertyService.cs
using RealEstateKayt.Models;

namespace RealEstateKayt.Services;

public class PropertyService
{
    private List<Property> _properties = new();

    public IEnumerable<Property> GetAll() => _properties;

    public void Add(Property property) => _properties.Add(property);

    // Search properties by address or description
    public IEnumerable<Property> Search(string term)
    {
        return _properties
            .Where(p =>
                (!string.IsNullOrWhiteSpace(term)) &&
                (p.Address.Contains(term, StringComparison.OrdinalIgnoreCase) ||
                 p.Description.Contains(term, StringComparison.OrdinalIgnoreCase)));
    }

    // Filter by availability
    public IEnumerable<Property> FilterAvailable(bool onlyAvailable)
    {
        return onlyAvailable ? _properties.Where(p => p.IsAvailable) : _properties;
    }

    // Sort by price
    public IEnumerable<Property> SortByPrice(bool descending = false)
    {
        return descending
            ? _properties.OrderByDescending(p => p.PricePerNight)
            : _properties.OrderBy(p => p.PricePerNight);
    }

    // Combined query: search + filter + sort
    public IEnumerable<Property> Query(string? searchTerm, bool onlyAvailable, bool sortDesc)
    {
        var result = _properties.AsEnumerable();

        if (!string.IsNullOrWhiteSpace(searchTerm))
            result = result.Where(p =>
                p.Address.Contains(searchTerm, StringComparison.OrdinalIgnoreCase) ||
                p.Description.Contains(searchTerm, StringComparison.OrdinalIgnoreCase));

        if (onlyAvailable)
            result = result.Where(p => p.IsAvailable);

        result = sortDesc
            ? result.OrderByDescending(p => p.PricePerNight)
            : result.OrderBy(p => p.PricePerNight);

        return result;
    }
}
